### PR TITLE
Add --config-file argument to use alternate config file at runtime

### LINKED ISSUE #1170: Specify an alternate config file in the command line
## Feature Request

It would be nice to specify an alternate config file from the command line.

### Use Case/Motivation

We already have an issue for specific overriding configuration values on the command line (#1157), but in many situations, it would be a lot cleaner and more useful for users to be able to specify a full config file.

This could lead to some privacy-oriented use cases like a portable jrnl running on a removable drive, with no trace of its configuration on the operating system's drive. It could also be useful for testing and troubleshooting.

### Example Usage

```
# New entry
jrnl --config-file g:\portableconfig.yaml

# Viewing last five entries
jrnl --config-file g:\portableconfig.yaml -5

# Order of arguments shouldn't matter
jrnl -5 --config-file g:\portableconfig.yaml

# Writing a new entry on a particular journal specified in the alternate config file
jrnl secondary --config-file g:\portableconfig.yaml
```

### Other information

Would be nice to provide an alias for this argument, like --cf.

Some thought would need to go into how compatible this should (or shouldn't) be with features that change the config, such as  #1068 (modify config) and #330 (create new journals).

### PR DESCRIPTION
**Pull request by samuelgregorovic**
Allows user to specify alternate config file for given session via `--config-file` argument. Fixes [#1170](https://github.com/jrnl-org/jrnl/issues/1170)

### CODE REVIEW

**Comment by wren** [comment]
Thank you for submitting this! We'll take a look at it on Saturday this week.

**Review by micahellison — CHANGES_REQUESTED** [review]
This looks great. Thanks for submitting this, and for adding in documentation along with the improvement.

Would you be willing to add a couple BDD tests to ensure support for multiple journals? I see that it's currently working, but would like to be sure that it stays working as people modify this in the future. Basically, it would confirm that jrnl can read and write to a non-default journal while using the alternative config file argument:
```
jrnl secondjournal --cf test_config.yaml new journal entry
jrnl secondjournal --cf test_config.yaml -1
```

File: docs/advanced.md [hunk 1]
```diff
@@ -85,6 +85,25 @@ jrnl --config-override display_format fancy --config-override linewrap 20 \
 
 ```
 
+### Using an alternate config
+
+You can specify an alternate configuration file fot the current instance of `jrnl` using `--config-file CONFIG_FILE_PATH` where
```
Comment (wren) [inline docs/advanced.md]:
```suggestion
You can specify an alternate configuration file for the current instance of `jrnl` using `--config-file CONFIG_FILE_PATH` where
```

**Comment by samuelgregorovic** [comment]
> This looks great. Thanks for submitting this, and for adding in documentation along with the improvement.
> 
> Would you be willing to add a couple BDD tests to ensure support for multiple journals? I see that it's currently working, but would like to be sure that it stays working as people modify this in the future. Basically, it would confirm that jrnl can read and write to a non-default journal while using the alternative config file argument:
> ```
> jrnl secondjournal --cf test_config.yaml new journal entry
> jrnl secondjournal --cf test_config.yaml -1
> ```

i dont really have experience with BDD but i can try to do it, if you won't suggest otherwisw

**Comment by micahellison** [comment]
> i dont really have experience with BDD but i can try to do it

Feel free to take a look at `\tests\bdd\features\multiple_journals.feature` on the develop branch. You should be able to use roughly the same syntax as the tests in there to add in this use case.

**Comment by wren** [comment]
@samuelgregorovic If you're up for it, please give it a shot! And if it doesn't work out or you need any help, just let us know and we can jump in.

**Comment by samuelgregorovic** [comment]
> @samuelgregorovic If you're up for it, please give it a shot! And if it doesn't work out or you need any help, just let us know and we can jump in.

deal!

**Comment by samuelgregorovic** [comment]
> > i dont really have experience with BDD but i can try to do it
> 
> Feel free to take a look at `\tests\bdd\features\multiple_journals.feature` on the develop branch. You should be able to use roughly the same syntax as the tests in there to add in this use case.

thanks for the heads up!

**Comment by samuelgregorovic** [comment]
@wren @micahellison latest commit contains the BDD tests as you wanted. could you please look at them if they satisfy your expectations? thanks

**Comment by wren** [comment]
@samuelgregorovic This is looking great. I think we didn't quite clarify one important point, though.

We currently have two BDD frameworks in place:
1. **Behave** (in `/features` folder)
  We're currently phasing out Behave, so we don't want to update or fix anything in the `/features` folder anymore.

2. **Pytest BDD** (in `/tests/bdd` folder)
  We recently moved over all of the tests from Behave into this in order to centralize all of our tests into Pytest (which we use for unit tests). Pytest BDD will be our BDD testing framework for the forseeable future.

Sorry for the misunderstanding! The work you've done here looks great, though. It seems like you took a current feature file, copied it over, and changed the tests to work for your tests. If you do the same thing, but start with the file mentioned above (`/tests/bdd/features/multiple_journals.feature`), then I think you'll be good to go.

The only bit of feedback I have is try to use existing config files if at all possible, rather than adding any new ones (they're kind of a pain to maintain, and we're considering getting rid of them in favor of Pytest fixtures in the future).

**Comment by wren** [comment]
Oh, also, there seems to be a merge conflict in `tests/test_parse_args.py`. I'm not sure what the conflict is, but you can resolve it if you merge the `develop` branch into your PR.

**Comment by samuelgregorovic** [comment]
@wren sorry for the misunderstanding. i will take a look

**Comment by wren** [comment]
@samuelgregorovic No worries!

Also, the develop branch had broken docs tests, which is affecting this PR. If you merge in the develop branch, it should fix that for you. Sorry!

**Comment by micahellison** [comment]
Hi there @samuelgregorovic, would you like any help with the pytest-bdd tests? If you're not working on them, we can add them in.

**Comment by samuelgregorovic** [comment]
> Hi there @samuelgregorovic, would you like any help with the pytest-bdd tests? If you're not working on them, we can add them in.

im sorry i was on a holiday and i left it unfinished. i will do it in the following days

**Comment by wren** [comment]
> > Hi there @samuelgregorovic, would you like any help with the pytest-bdd tests? If you're not working on them, we can add them in.
> 
> im sorry i was on a holiday and i left it unfinished. i will do it in the following days

No worries! This is all volunteer work, and it's always absolutely fine too take your time. Thanks for the response, and please let us know whenever you need any help.

**Comment by micahellison** [comment]
Hi there @samuelgregorovic -- I started the process of migrating the tests to pytest/bdd and pushed it to this PR, but partway through the process, I found a bug related to the upgrade for older configs. Basically, if you reference the `multiple.yaml` file in the config using `--cf`, it will upgrade it and save over your *main* config file.

If you want to see it in action, backup your jrnl config (or commit it in its own git repo) before running this from the jrnl root:
```
poetry run jrnl --cf tests/data/configs/multiple.yaml this is a test
```

**Comment by micahellison** [comment]
I fixed that issue with overwriting the default config, then updated the rest of the tests for pytest-bdd. Had to deal with some Python 3.7 issues and a weird timeout with pa11y that somehow resolved after I cleaned up a couple small errors in the documentation. I think this is ready for review now.

**Review by micahellison — DISMISSED** [review]

File: tests/lib/given_steps.py [hunk 2]
```diff
@@ -93,7 +93,10 @@ def we_use_the_config(config_file, temp_dir, working_dir):
     # Copy all of the journals over
     journal_source = os.path.join(working_dir, "data", "journals")
     journal_dest = os.path.join(temp_dir.name, "features", "journals")
-    shutil.copytree(journal_source, journal_dest)
+    try:
```
Comment (wren) [inline tests/lib/given_steps.py]: Do we still need this?

**Review by wren — APPROVED** [review]
🎸

**Pull request merged**